stats/registry.go

```diff
@@ -0,0 +1,91 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2021 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package stats
+
+import (
+	"context"
+	"fmt"
+	"sync"
+)
+
+// Registry is what can create metrics
+// TODO maybe rename the whole package to metrics so we have `metrics.Registry` ?
+type Registry struct {
+	metrics map[string]*Metric
+	sink    chan<- SampleContainer
+	l       sync.RWMutex
+	// TODO maybe "seal" it after we no longer want to get new metrics?
+}
+
+func NewRegistry(sink chan<- SampleContainer) *Registry {
+	return &Registry{
+		metrics: make(map[string]*Metric),
+		sink:    sink,
+	}
+}
+
+// NewMetric returns new metric registered to this registry
+// TODO have multiple versions returning specific metric types when we have such things
+func (r *Registry) NewMetric(name string, typ MetricType, t ...ValueType) (*Metric, error) {
+	r.l.Lock()
+	defer r.l.Unlock()
+	oldMetric, ok := r.metrics[name]
+
+	if !ok {
+		m := New(name, typ, t...)
+		m.r = r
+		r.metrics[name] = m
+		return m, nil
+	}
+	// Name is checked. TODO check Contains as well
+	if oldMetric.Type != typ {
+		return nil, fmt.Errorf("Metric `%s` already exist but with type %s", name, oldMetric.Type)
+	}
+	return oldMetric, nil
+}
+
+func (r *Registry) MustNewMetric(name string, typ MetricType, t ...ValueType) *Metric {
+	m, err := r.NewMetric(name, typ, t...)
+	if err != nil {
+		panic(err)
+	}
+	return m
+}
+
+// PushContainer emits a contained through the registry's sink
+// TODO drop this somehow
+func (r *Registry) PushContainer(ctx context.Context, container SampleContainer) {
+	PushIfNotDone(ctx, r.sink, container)
+}
+
```

Comment:
A lot of the actual design problems come from the fact we have containers and even more because of custom containers such as HTTPTrail.

I don't think that just removing the later is feasible at all without so much rework that it's totally a different PR, even if we wanted to, so some workaround will need to be found. 

The predominant problem is that while I can make certain that the only way the type Sample can be emitted is through the registry, by just not having a different way to emit a metric through the registry but through the metric it generates (see Sample's Push method which, probably should be an Emit on the Metric :thinking: ), this more or less doesn't work with SampleContainers and especially with HTTPTrail as that one is specifically useful in the cloud output, where instead of looking at it's Samples we just look at it ... properly typed fields. 

I do think we should postpone fixing that for a later PR,hopefully still in the v0.33.0 cycle. I have hopes that if we introduce separate Metric/Sample types for Gauge, Counter, Rate and Trend that will make some this problems go away :shrug: